### PR TITLE
[functorch] disable C++ Function under functorch transforms

### LINKED ISSUE #102720: vmap breaks gradient on custom c++ functions
### 🐛 Describe the bug

When calling vmap with custom c++ functions, the gradients are broken:
I'm basing this on the [safetanh from torchrl](https://github.com/pytorch/rl/blob/fa45472da29cc1d07b3ae57bdfe094207deeff13/torchrl/csrc/utils.h#L13-L31)

```python
>>> import torch
>>> from torch import vmap
>>> from torchrl._torchrl import safeatanh, safetanh
>>> safetanh(torch.zeros(3, requires_grad=True), 1e-6)
tensor([0., 0., 0.], grad_fn=<CppNode<SafeTanh>>)
>>> vmap(safetanh, (0, None))(torch.zeros(3, requires_grad=True), 1e-6)
tensor([0., 0., 0.])
>>> # for ref
>>> vmap(torch.tanh)(torch.zeros(3, requires_grad=True))
tensor([0., 0., 0.], grad_fn=<TanhBackward0>)
```

I guess I should fix the c++ code in some way, but maybe we should raise an exception somewhere here?

cc @ezyang @gchanan @zou3519 @albanD @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @Chillee @samdow @kshitij12345 @janeyx99 

### Versions

2.1.0.dev20230525

## Edit

I followed https://pytorch.org/docs/master/notes/extending.func.html and I managed to make it work in python.
The python error was very explicit, that really helped.

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes https://github.com/pytorch/pytorch/issues/102720

### CODE REVIEW

File: aten/src/ATen/functorch/DynamicLayer.cpp [hunk 1]
```diff
@@ -99,6 +99,13 @@ class FuncTorchTLS : public FuncTorchTLSBase {
     return 0;
   }
 
+  void checkSupportsCppAutogradFunction() const override {
+    TORCH_CHECK(
+        dynamicLayerStack.empty(),
+        "functorch functions (vmap, grad, vjp, etc.) incorrectly used with ",
+        "C++ Function. ");
```
Comment (zou3519) [inline aten/src/ATen/functorch/DynamicLayer.cpp:106]: Nit: "cannot use C++ torch::autograd::Function with functorch transforms (vmap, grad, vjp, etc)".

File: torch/csrc/autograd/custom_function.h [hunk 2]
```diff
@@ -263,6 +263,14 @@ template <class T>
 template <typename X, typename... Args>
 auto Function<T>::apply(Args&&... args)
     -> std::enable_if_t<std::is_same<X, T>::value, forward_t<X, Args...>> {
+  const auto& functorch_tls = at::functorch::functorchTLSAccessor();
+  if (functorch_tls) {
+    // Function support for functorch is handled in Python.
+    // Here we are dealing with a (C++) Function, which is not supported.
+    // Let's raise an error instead of being silently incorrect.
+    functorch_tls->checkSupportsCppAutogradFunction();
```
[the remark this answers is not in the document]
Comment (zou3519) [reply torch/csrc/autograd/custom_function.h:271]: Nope, we want this for all functorch transforms.

If we vmap over the torch::autograd::Function's forward, it turns out that the backward pass isn't preserved (because the torch::autograd::Function gets invoked on BatchedTensors that cannot have grad_fn).

**Review by zou3519 — APPROVED** [review]
LGTM! Minor nit on error message
